### PR TITLE
Partial aggregation + misc improvements

### PR DESCRIPTION
**Pull request by pkozikow**

### CODE REVIEW

File: pigpen-cascading/src/main/clojure/pigpen/cascading.clj [hunk 1]
```diff
@@ -66,39 +61,102 @@
...
-                                          (:group-all cogroup-opts)
-                                          (:join-nils cogroup-opts)
-                                          (:join-requirements cogroup-opts)
-                                          udf
-                                          key-separate-from-value)
-                     :join (JoinBuffer. (first inits) (first funcs) fields (:all-args cogroup-opts)))]
-        (update-in flowdef [:pipes pipe] #(Every. % buffer Fields/RESULTS)))
-      (update-in flowdef [:pipes pipe] #(Each. % (PigPenFunction. (first inits) (first funcs) fields) Fields/RESULTS)))))
+        cogroup-opts (get-in flowdef [:cogroup-opts pipe-id])]
+    (if (nil? cogroup-opts)
```
Comment (mbossenbroek) [inline pigpen-cascading/src/main/clojure/pigpen/cascading.clj]: `cond` might be nicer than the nested `if`s

File: pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java [hunk 2]
```diff
@@ -13,6 +13,8 @@
 
 public class OperationUtil {
 
+  public static final String SENTINEL_VALUE = "39c63d213f5cba3c7";
```
Comment (mbossenbroek) [inline pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java]: Interesting choice for a sentinel value. I generally use namespace qualified keywords - would a similar approach not work here?

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java [hunk 3]
```diff
@@ -9,30 +9,40 @@
...
 /**
- * Sequence that allows only one iteration, but is immune to
- * memory problems resulting from holding onto the head.
+ * This sequence behaves like a normal iterator-seq up to a certain size.
+ * If the max size is exceeded, it will only allow one iteration. The objective
+ * is to become immune to memory problems resulting from holding onto the head for
+ * large collections but still support operations such as count, max, etc.
  */
 public class SingleIterationSeq extends ASeq {
+  private static int MAX_SIZE = 1000;
```
Comment (mbossenbroek) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: Just curious, why 1000?

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java [hunk 4]
```diff
@@ -53,16 +63,24 @@ public Object first() {
   }
 
   public ISeq next() {
-    if (state.isRealized == false) {
+    if (state.rest == state) {
```
Comment (mbossenbroek) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: Will there be multi-threaded access to this? Does the synchronization add any noticeable overhead?

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java (hunk 3 above)
Comment (pkozikow) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: This was an arbitrary nice round number that sounded reasonable (i.e. large enough for most use cases that require multiple passes, but small enough not to cause memory issues). I'm open to suggestions, and we could make this configurable.

File: pigpen-cascading/src/main/clojure/pigpen/cascading.clj (hunk 1 above)
Comment (pkozikow) [inline pigpen-cascading/src/main/clojure/pigpen/cascading.clj]: Indeed, I'll change it.

File: pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java (hunk 2 above)
Comment (pkozikow) [inline pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java]: I didn't put much thought into it. The thing is that it is easier to use a type that Cascading knows how to serialize natively. I could use the serialized version of a namespace qualified keyword though. What would be the pros and cons?

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java (hunk 3 above)
Comment (mbossenbroek) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: What cases exercise the multiple passes? I don't fully get when that would be used.

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java (hunk 4 above)
Comment (pkozikow) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: I was wondering the same. The code is an adaptation of clojure's IteratorSeq, which is what I was using originally but caused OOMEs. If you think it's safe to assume PigPen users will never want to use multi-threaded access from their UDFs, I'm happy to remove the synchronization.

File: pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java (hunk 2 above)
Comment (mbossenbroek) [inline pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java]: Maybe "pigpen.cascading.OperationUtil/SENTINEL_VALUE"? The idea is to find a value that the user would _never_ use. It's entirely a gut feeling, but it seems like random numbers could occur in user data, but this namespace should never be in their code. Something random is almost less avoidable than something very specific, but unlikely to be in their data.

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java (hunk 4 above)
Comment (mbossenbroek) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: That's a good point - I guess if they use a `pmap`, we could run into issues. Let's leave it in.

File: pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java (hunk 3 above)
Comment (pkozikow) [inline pigpen-cascading/src/main/java/pigpen/cascading/SingleIterationSeq.java]: For one thing a bunch of tests from functional_test.clj fail without this. I missed that fact when I introduced the original version. But in general I can imagine users wanting to perform several unrelated operations to the same seq, and then combine them in some way in their UDF.

File: pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java (hunk 2 above)
Comment (pkozikow) [inline pigpen-cascading/src/main/java/pigpen/cascading/OperationUtil.java]: Fair enough. I was also trying to use something that is not too large, but correctness guarantee comes first.

**Pull request merged**